# **Ejercicio 4**
# **Single Cell Perturbations**
##  **Modelos de Regresión**
Considere el conjunto de datos `Open Problems– Single-Cell Perturbations`. Implemente la versión de regresión de cada uno de los modelos estudiados en clases, esto es KNN y Regresión Lineal en el conjunto de datos suministrado. Construir una tabla de error con las métricas usuales de regresión, MAPE, MAE, RMSE, MSE, R2 (ver Table 2). Utilice la métrica Mean Rowwise Root Mean Squared Error (MRRMSE) en la evaluación y validación, para seleccionar el mejor modelo de regresión.



**Cuadro 2: Modelo de regresión para velocidad del viento**

| **Modelo**            | **MAPE** | **MAE** | **RMSE** | **MSE** | **R2** |
|-----------------------|----------|---------|----------|---------|--------|
| K-NN                  | ...      | ...      | ...       | ...      | ...     |
| Linear Regression     | ...      | ...      | ...       | ...      | ...     |



### **Métrica asignada**

Esta métrica corresponde a la sugerida por el enunciado, además esta mencionada en la información de la base de datos.

In [ ]:
def MRRMSE(y_real, y_pred):
    rowwise_rmse = np.sqrt(np.mean((y_real - y_pred) ** 2, axis=1))
    return np.mean(rowwise_rmse)


La métrica **MRRMSE** (Mean Row-wise Root Mean Square Error) mide el error promedio entre los valores predichos y los reales por fila, lo que permite evaluar la precisión de predicciones en un conjunto de datos estructurados en filas, como en matrices de expresión génica. Primero calcula el **RMSE** por fila, para luego promediar estos valores, destacando su capacidad de capturar errores individuales por observación. En este contexto, se espera que el **MRRMSE** sea lo más bajo posible, lo que indicaría que las predicciones están muy cercanas a los valores reales en cada fila, lo que refleja un buen ajuste del modelo.

## **Datos**

Después de haber realizado un análisis descriptivo de la base de datos original (detallado en el ejercicio 2), finalmente se ha obtenido una base de datos codificada y preparada, la cual ha sido almacenada en un archivo CSV.

In [ ]:
de_train_final = pd.read_csv('C:/Users/kamac/OneDrive/Desktop/MachineLearningUN/train.csv')

In [ ]:
de_train_final

cell_type_B cells  cell_type_Myeloid cells  cell_type_NK cells  \
0                  0.0                      0.0                 1.0   
1                  0.0                      0.0                 0.0   
2                  0.0                      0.0                 0.0   
3                  0.0                      0.0                 0.0   
4                  0.0                      0.0                 1.0   
..                 ...                      ...                 ...   
609                0.0                      0.0                 0.0   
610                0.0                      0.0                 1.0   
611                0.0                      0.0                 0.0   
612                0.0                      0.0                 0.0   
613                0.0                      0.0                 0.0   

     cell_type_T cells CD4+  cell_type_T cells CD8+  \
0                       0.0                     0.0   
1                       1.0                     0.0   
2                       0.0                     1.0   
3                       0.0                     0.0   
4                       0.0                     0.0   
..                      ...                     ...   
609                     0.0                     0.0   
610                     0.0                     0.0   
611                     1.0                     0.0   
612                     0.0                     1.0   
613                     0.0                     0.0   

     cell_type_T regulatory cells  \
0                             0.0   
1                             0.0   
2                             0.0   
3                             1.0   
4                             0.0   
..                            ...   
609                           1.0   
610                           0.0   
611                           0.0   
612                           0.0   
613                           1.0   

     sm_name_5-(9-Isopropyl-8-methyl-2-morpholino-9H-purin-6-yl)pyrimidin-2-amine  \
0                                                  0.0                              
1                                                  0.0                              
2                                                  0.0                              
3                                                  0.0                              
4                                                  0.0                              
..                                                 ...                              
609                                                0.0                              
610                                                0.0                              
611                                                0.0                              
612                                                0.0                              
613                                                0.0                              

     sm_name_ABT-199 (GDC-0199)  sm_name_ABT737  \
0                           0.0             0.0   
1                           0.0             0.0   
2                           0.0             0.0   
3                           0.0             0.0   
4                           0.0             0.0   
..                          ...             ...   
609                         0.0             0.0   
610                         0.0             0.0   
611                         0.0             0.0   
612                         0.0             0.0   
613                         0.0             0.0   

     sm_name_AMD-070 (hydrochloride)  ...      ZUP1      ZW10    ZWILCH  \
0                                0.0  ... -0.227781 -0.010752 -0.023881   
1                                0.0  ... -0.494985 -0.303419  0.304955   
2                                0.0  ... -0.119422 -0.033608 -0.153123   
3                                0.0  ...  0.451679  0.704643  0.015468   
4                                0.0 

Nótese que la base de datos actual cuenta con **614** observaciones y **18363** columnas

Ahora, se divide el DataFrame `de_train_final` en dos partes: las variables predictoras (almacenadas en `x`) y las variables objetivo (almacenadas en `y`). Específicamente, las primeras 152 columnas de la base de datos se asignan a `x`, que representará los predictores o características, mientras que las columnas a partir de la posición 152 en adelante se asignan a `y`, que representan las respuestas o la variable objetivo.

In [ ]:
X = de_train_final.iloc[:,:152]
y = de_train_final.iloc[:,152:]

In [ ]:
X

cell_type_B cells  cell_type_Myeloid cells  cell_type_NK cells  \
0                  0.0                      0.0                 1.0   
1                  0.0                      0.0                 0.0   
2                  0.0                      0.0                 0.0   
3                  0.0                      0.0                 0.0   
4                  0.0                      0.0                 1.0   
..                 ...                      ...                 ...   
609                0.0                      0.0                 0.0   
610                0.0                      0.0                 1.0   
611                0.0                      0.0                 0.0   
612                0.0                      0.0                 0.0   
613                0.0                      0.0                 0.0   

     cell_type_T cells CD4+  cell_type_T cells CD8+  \
0                       0.0                     0.0   
1                       1.0                     0.0   
2                       0.0                     1.0   
3                       0.0                     0.0   
4                       0.0                     0.0   
..                      ...                     ...   
609                     0.0                     0.0   
610                     0.0                     0.0   
611                     1.0                     0.0   
612                     0.0                     1.0   
613                     0.0                     0.0   

     cell_type_T regulatory cells  \
0                             0.0   
1                             0.0   
2                             0.0   
3                             1.0   
4                             0.0   
..                            ...   
609                           1.0   
610                           0.0   
611                           0.0   
612                           0.0   
613                           1.0   

     sm_name_5-(9-Isopropyl-8-methyl-2-morpholino-9H-purin-6-yl)pyrimidin-2-amine  \
0                                                  0.0                              
1                                                  0.0                              
2                                                  0.0                              
3                                                  0.0                              
4                                                  0.0                              
..                                                 ...                              
609                                                0.0                              
610                                                0.0                              
611                                                0.0                              
612                                                0.0                              
613                                                0.0                              

     sm_name_ABT-199 (GDC-0199)  sm_name_ABT737  \
0                           0.0             0.0   
1                           0.0             0.0   
2                           0.0             0.0   
3                           0.0             0.0   
4                           0.0             0.0   
..                          ...             ...   
609                         0.0             0.0   
610                         0.0             0.0   
611                         0.0             0.0   
612                         0.0             0.0   
613                         0.0             0.0   

     sm_name_AMD-070 (hydrochloride)  ...  sm_name_Tivozanib  \
0                                0.0  ...                0.0   
1                                0.0  ...                0.0   
2                                0.0  ...                0.0   
3                                0.0  ...                0.0   
4                                0.0  ...                0.0   
..                          

In [ ]:
y

A1BG  A1BG-AS1       A2M   A2M-AS1     A2MP1    A4GALT      AAAS  \
0    0.104720 -0.077524 -1.625596 -0.144545  0.143555  0.073229 -0.016823   
1    0.915953 -0.884380  0.371834 -0.081677 -0.498266  0.203559  0.604656   
2   -0.387721 -0.305378  0.567777  0.303895 -0.022653 -0.480681  0.467144   
3    0.232893  0.129029  0.336897  0.486946  0.767661  0.718590 -0.162145   
4    4.290652 -0.063864 -0.017443 -0.541154  0.570982  2.022829  0.600011   
..        ...       ...       ...       ...       ...       ...       ...   
609 -0.014372 -0.122464 -0.456366 -0.147894 -0.545382 -0.544709  0.282458   
610 -0.455549  0.188181  0.595734 -0.100299  0.786192  0.090954  0.169523   
611  0.338168 -0.109079  0.270182 -0.436586 -0.069476 -0.061539  0.002818   
612  0.101138 -0.409724 -0.606292 -0.071300 -0.001789 -0.706087 -0.620919   
613 -0.757116  0.085910 -0.730025 -1.367801 -0.695944 -0.724985  0.121436   

         AACS     AAGAB      AAK1  ...      ZUP1      ZW10    ZWILCH  \
0    0.101717 -0.005153  1.043629  ... -0.227781 -0.010752 -0.023881   
1    0.498592 -0.317184  0.375550  ... -0.494985 -0.303419  0.304955   
2   -0.293205 -0.005098  0.214918  ... -0.119422 -0.033608 -0.153123   
3    0.157206 -3.654218 -0.212402  ...  0.451679  0.704643  0.015468   
4    1.231275  0.236739  0.338703  ...  0.758474  0.510762  0.607401   
..        ...       ...       ...  ...       ...       ...       ...   
609 -0.431359 -0.364961  0.043123  ... -0.549987 -2.200925  0.359806   
610  0.428297  0.106553  0.435088  ... -1.236905  0.003854 -0.197569   
611 -0.027167 -0.383696  0.226289  ...  0.077579 -1.101637  0.457201   
612 -1.485381  0.059303 -0.032584  ...  0.005951 -0.893093 -1.003029   
613 -0.785661 -0.240842 -0.139620  ...  0.232343 -2.247816 -0.346036   

        ZWINT      ZXDA      ZXDB      ZXDC    ZYG11B       ZYX     ZZEF1  
0    0.674536 -0.453068  0.005164 -0.094959  0.034127  0.221377  0.368755  
1   -0.333905 -0.315516 -0.369626 -0.095079  0.704780  1.096702 -0.869887  
2    0.183597 -0.555678 -1.494789 -0.213550  0.415768  0.078439 -0.259365  
3   -0.103868  0.865027  0.189114  0.224700 -0.048233  0.216139 -0.085024  
4   -0.123059  0.214366  0.487838 -0.819775  0.112365 -0.122193  0.676629  
..        ...       ...       ...       ...       ...       ...       ...  
609  1.073983  0.356939 -0.029603 -0.528817  0.105138  0.491015 -0.979951  
610 -0.175307  0.101391  1.028394  0.034144 -0.231642  1.023994 -0.064760  
611  0.535184 -0.198404 -0.005004  0.552810 -0.209077  0.389751 -0.337082  
612 -0.080367 -0.076604  0.024849  0.012862 -0.029684  0.005506 -1.733112  
613 -0.919567 -1.131372 -0.120252 -0.064537 -0.603280 -0.098041 -0.750681  

[614 rows x 18211 columns]

Antes de aplicar los modelos se divide el conjunto de datos en dos partes: entrenamiento y prueba. Con la función `train_test_split`, las variables `X` y `y` se dividen en `X_train` y `X_test` (para las características) y en `y_train` y `y_test` (para la variable objetivo). Además, se está crea un "scoring" personalizado utilizando la función make_scorer, con la métrica MRRMSE (definida anteriormente), donde greater_is_better=False indica que se considera mejor un valor más bajo para la métrica, es decir, cuanto menor sea el error, mejor será el modelo.

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 42)
scoring = make_scorer(MRRMSE, greater_is_better = False) # Indica que un valor más bajo de la métrica MRRMSE es mejor

En este código, se entrena un modelo **KNN Regressor** usando un **pipeline** con estandarización (`StandardScaler()`). Se realiza una búsqueda de hiperparámetros con **GridSearchCV** utilizando un conjunto de parámetros que incluyen el número de vecinos (`n_neighbors`), la métrica de distancia (`metric`), el método de ponderación (`weights`), y el parámetro `p` para la métrica Minkowski. Se utiliza **KFold** con 10 divisiones y se evalúa el modelo mediante la métrica **MRRMSE**.

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 11)

pipeline_kr = Pipeline([('scaler', MinMaxScaler()),('knr', KNeighborsRegressor())])

kfo = KFold(n_splits = 10, shuffle = True, random_state = 11)

param_grid_kr = {
    'knr__n_neighbors': list(range(1, 200)),
    'knr__metric': ['minkowski', 'euclidean', 'manhattan'],
    'knr__p': [1, 2, 3],
    'knr__weights': ['distance']
}


In [ ]:
grid_knr = GridSearchCV(pipeline_kr, param_grid_kr, cv = kfo, scoring = scoring, verbose = 3, n_jobs = -1)
grid_knr.fit(X_train, y_train)

modelo = grid_knr.best_estimator_

y_pred_test = modelo.predict(X_test)
mrrmse_test = MRRMSE(y_test, y_pred_test)

print(f'Mejores parámetros: {grid_knr.best_params_}')
print(f'MRRMSE en el conjunto de entrenamiento: {grid_knr.best_score_}')
print(f'MRRMSE en el conjunto de prueba: {mrrmse_test}')

Fitting 10 folds for each of 1791 candidates, totalling 17910 fits
Mejores parámetros: {'knr__metric': 'minkowski', 'knr__n_neighbors': 125, 'knr__p': 3, 'knr__weights': 'distance'}
MRRMSE en el conjunto de entrenamiento: -1.2264458068125894
MRRMSE en el conjunto de prueba: 1.470800131934799


Los resultados muestran que el modelo de KNN- Regressor, con sus mejores parámetros encontrados (métrica Minkowski, 125 vecinos y pesos por distancia), tiene un desempeño deficiente, especialmente en el conjunto de entrenamiento, donde el MRRMSE es negativo (-1.22), lo que sugiere sobreajuste o problemas con los datos. En el conjunto de prueba, el MRRMSE es 1.47, lo que indica que el modelo sigue teniendo dificultades para predecir con precisión, aunque con un mejor desempeño en comparación con el entrenamiento. Esto refleja que el modelo no está generalizando bien a los nuevos datos.

### **Modelo de regresión lineal**

Se entrena un modelo de **regresión lineal** utilizando un **pipeline** que incluye la estandarización de los datos con `StandardScaler()`. No se definen hiperparámetros, ya que la regresión lineal no tiene parámetros ajustables y se evalúa con la métrica **MRRMSE**. .

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 11)

pipeline_lr = Pipeline([('scaler', StandardScaler()), ('lr', LinearRegression())])

pipeline_lr.fit(X_train, y_train)

y_pred_train_lr = pipeline_lr.predict(X_train)
mrrmse_train_lr = MRRMSE(y_train, y_pred_train_lr)

y_pred_lr = pipeline_lr.predict(X_test)
mrrmse_lr = MRRMSE(y_test, y_pred_lr)


print(f'MRRMSE en el conjunto de entrenamiento: {mrrmse_train_lr}')
print(f'MRRMSE en el conjunto de prueba: {mrrmse_lr}')


MRRMSE en el conjunto de entrenamiento: 0.9867218722252062
MRRMSE en el conjunto de prueba: 1.5387031905624788


Los resultados muestran que el MRRMSE en el conjunto de entrenamiento es de 0.98, lo que indica un buen ajuste. Sin embargo, el MRRMSE en el conjunto de prueba es mayor (1.53), lo que sugiere que el modelo tiene un peor rendimiento al predecir datos nuevos. Esto podría ser indicativo de sobreajuste, ya que el modelo se ajusta bien al conjunto de entrenamiento, pero no generaliza tan bien en el conjunto de prueba.

### **Cuadro 2**

Finalmente se realizan predicciones con los modelos **K-NN** y **regresión lineal**, luego se evalúan usando diversas métricas de error, como **MAPE** (Error absoluto medio en porcentaje), **MAE** (Error absoluto medio), **RMSE** (Raíz cuadrada del error cuadrático medio), **MSE** (Error cuadrático medio), y **R2** (Coeficiente de determinación).

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 10)
knr_predict = grid_knr.predict(X_test)
lr_predict = pipeline_lr.predict(X_test)

tabla2 = {
    'Modelo': ['K-NN', 'Regresión lineal'],
    'MAPE': [  # Error absoluto medio en porcentaje
        mean_absolute_percentage_error(y_test, knr_predict),
        mean_absolute_percentage_error(y_test, lr_predict)
    ],
    'MAE': [  # Error absoluto medio
        mean_absolute_error(y_test, knr_predict),
        mean_absolute_error(y_test, lr_predict)
    ],
    'RMSE': [  # Raíz cuadrada del error cuadrático medio
        np.sqrt(mean_squared_error(y_test, knr_predict)),
        np.sqrt(mean_squared_error(y_test, lr_predict))
    ],
    'MSE': [  # Error cuadrático medio
        mean_squared_error(y_test, knr_predict),
        mean_squared_error(y_test, lr_predict)
    ],
    'R2': [  # Coeficiente de determinación
        r2_score(y_test, knr_predict),
        r2_score(y_test, lr_predict)
    ]
}

cuadro2 = pd.DataFrame(tabla2)
cuadro2


Modelo       MAPE       MAE      RMSE       MSE        R2
0              K-NN   1.060034  0.150213  1.006208  1.012454  0.791933
1  Regresión lineal  16.845013  0.735129  1.656074  2.742582  0.200860

En la tabla de métricas, el coeficiente de determinación $R^2$ mide qué tan bien el modelo explica la variabilidad de los datos. Un valor de  $R^2$ cercano a 1 indica un buen ajuste del modelo. En este caso, el modelo K-NN tiene un $R^2$ de 0.79, lo que significa que el 79% de la variabilidad de los datos es explicada por el modelo, lo cual es un buen resultado. Por otro lado, la regresión lineal tiene un $R^2$  de 0.20, lo que indica que solo el 20% de la variabilidad en los datos se explica por este modelo, lo que sugiere un ajuste mucho peor en comparación con K-NN.